File: samples/CameraAccessAndroid/app/src/test/java/com/meta/wearable/dat/externalsampleapps/cameraaccess/conference/ConferenceModeTests.kt
package com.meta.wearable.dat.externalsampleapps.cameraaccess.conference

import com.meta.wearable.dat.externalsampleapps.cameraaccess.openclaw.ToolDeclarations
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class ConferenceModeTests {

    @Test
    fun conferencePromptUsesFallbackWhenDisabled() {
        val fallback = "custom prompt"

        assertEquals(
            fallback,
            ConferencePrompts.activeSystemInstruction(
                conferenceModeEnabled = false,
                fallbackPrompt = fallback,
            ),
        )
    }

    @Test
    fun conferencePromptUsesRelationshipOpsWhenEnabled() {
        assertEquals(
            ConferencePrompts.relationshipOps,
            ConferencePrompts.activeSystemInstruction(
                conferenceModeEnabled = true,
                fallbackPrompt = "custom",
            ),
        )
    }

    @Test
    fun toolDeclarationsExcludeExtractEntityWhenConferenceModeDisabled() {
        val names = ToolDeclarations.allDeclarations(conferenceModeEnabled = false).map { it.name }

        assertEquals(listOf(ToolDeclarations.EXECUTE_NAME), names)
    }

    @Test
    fun toolDeclarationsIncludeExtractEntityWhenConferenceModeEnabled() {
        val names = ToolDeclarations.allDeclarations(conferenceModeEnabled = true).map { it.name }

        assertEquals(
            listOf(ToolDeclarations.EXECUTE_NAME, ToolDeclarations.EXTRACT_ENTITY_NAME),
            names,
        )
    }

    @Test
    fun processorAcceptsHighConfidenceExtraction() {
        val processor = makeProcessor()

        val result = processor.handle(
            args = mapOf(
                "name" to "Sarah Chen",
                "company" to "Acme AI",
                "role" to "VP Engineering",
                "source_type" to "badge",
                "confidence" to 0.92,
            ),
            nowMs = 0L,
        )

        assertTrue(result is ConferenceExtractionHandlingResult.Accepted)
        val extraction = (result as ConferenceExtractionHandlingResult.Accepted).extraction
        assertEquals("Sarah Chen", extraction.name)
        assertEquals(ConferenceSourceType.BADGE, extraction.sourceType)
        assertEquals(ConferenceExtractionDisposition.ACCEPTED, extraction.disposition)
    }

    @Test
    fun processorReturnsReviewForMidConfidenceExtraction() {
        val processor = makeProcessor()

        val result = processor.handle(
            args = mapOf(
                "name" to "Taylor Reed",
                "source_type" to "card",
                "confidence" to 0.61,
            ),
            nowMs = 0L,
        )

        assertTrue(result is ConferenceExtractionHandlingResult.Review)
        val extraction = (result as ConferenceExtractionHandlingResult.Review).extraction
        assertEquals(ConferenceExtractionDisposition.REVIEW, extraction.disposition)
        assertEquals(ConferenceSourceType.CARD, extraction.sourceType)
    }

    @Test
    fun processorIgnoresLowConfidenceExtraction() {
        val processor = makeProcessor()

        val result = processor.handle(
            args = mapOf(
                "name" to "Jamie Park",
                "source_type" to "booth",
                "confidence" to 0.32,
            ),
            nowMs = 0L,
        )

        assertTrue(result is ConferenceExtractionHandlingResult.IgnoredLowConfidence)
        val confidence = (result as ConferenceExtractionHandlingResult.IgnoredLowConfidence).confidence
        assertEquals(0.32, confidence, 0.0001)
    }

    @Test
    fun processorSuppressesDuplicatesWithinCooldown() {
        val processor = makeProcessor()
        val args = mapOf(
            "name" to "Morgan Lee",
            "company" to "OpenClaw",
            "source_type" to "badge",
            "confidence" to 0.88,
        )

        val firstResult = processor.handle(args = args, nowMs = 0L)
        assertTrue(firstResult is ConferenceExtractionHandlingResult.Accepted)

        val secondResult = processor.handle(args = args, nowMs = 5_000L)
        assertEquals(ConferenceExtractionHandlingResult.IgnoredDuplicate, secondResult)
    }

    private fun makeProcessor(): ConferenceExtractionProcessor {
        return ConferenceExtractionProcessor(
            ConferenceModeConfig(
                enabled = true,
                acceptedConfidenceMin = 0.70,
                reviewConfidenceMin = 0.50,
                duplicateCooldownMs = 10_000L,
            ),
        )
    }
}

File: samples/CameraAccess/CameraAccessTests/ConferenceModeTests.swift
import Foundation
import XCTest

@testable import CameraAccess

final class ConferenceModeTests: XCTestCase {

  override func tearDown() {
    SettingsManager.shared.resetAll()
    super.tearDown()
  }

  func testConferencePromptUsesFallbackWhenDisabled() {
    let fallback = "custom prompt"

    XCTAssertEqual(
      ConferencePrompts.activeSystemInstruction(conferenceModeEnabled: false, fallbackPrompt: fallback),
      fallback
    )
  }

  func testConferencePromptUsesRelationshipOpsWhenEnabled() {
    XCTAssertEqual(
      ConferencePrompts.activeSystemInstruction(conferenceModeEnabled: true, fallbackPrompt: "custom"),
      ConferencePrompts.relationshipOps
    )
  }

  func testToolDeclarationsExcludeExtractEntityWhenConferenceModeDisabled() {
    let names = ToolDeclarations.allDeclarations(conferenceModeEnabled: false).compactMap {
      $0["name"] as? String
    }

    XCTAssertEqual(names, [ToolDeclarations.executeName])
  }

  func testToolDeclarationsIncludeExtractEntityWhenConferenceModeEnabled() {
    let names = ToolDeclarations.allDeclarations(conferenceModeEnabled: true).compactMap {
      $0["name"] as? String
    }

    XCTAssertEqual(names, [ToolDeclarations.executeName, ToolDeclarations.extractEntityName])
  }

  func testProcessorAcceptsHighConfidenceExtraction() {
    var processor = makeProcessor()

    let result = processor.handle(
      args: [
        "name": "Sarah Chen",
        "company": "Acme AI",
        "role": "VP Engineering",
        "source_type": "badge",
        "confidence": 0.92
      ],
      now: Date(timeIntervalSince1970: 0)
    )

    guard case .accepted(let extraction) = result else {
      return XCTFail("Expected accepted extraction, got \(result)")
    }

    XCTAssertEqual(extraction.name, "Sarah Chen")
    XCTAssertEqual(extraction.company, "Acme AI")
    XCTAssertEqual(extraction.sourceType, .badge)
    XCTAssertEqual(extraction.disposition, .accepted)
  }

  func testProcessorReturnsReviewForMidConfidenceExtraction() {
    var processor = makeProcessor()

    let result = processor.handle(
      args: [
        "name": "Taylor Reed",
        "source_type": "card",
        "confidence": 0.61
      ],
      now: Date(timeIntervalSince1970: 0)
    )

    guard case .review(let extraction) = result else {
      return XCTFail("Expected review extraction, got \(result)")
    }

    XCTAssertEqual(extraction.disposition, .review)
    XCTAssertEqual(extraction.sourceType, .card)
  }

  func testProcessorIgnoresLowConfidenceExtraction() {
    var processor = makeProcessor()

    let result = processor.handle(
      args: [
        "name": "Jamie Park",
        "source_type": "booth",
        "confidence": 0.32
      ],
      now: Date(timeIntervalSince1970: 0)
    )

    guard case .ignoredLowConfidence(let confidence) = result else {
      return XCTFail("Expected low-confidence ignore, got \(result)")
    }

    XCTAssertEqual(confidence, 0.32, accuracy: 0.0001)
  }

  func testProcessorSuppressesDuplicatesWithinCooldown() {
    var processor = makeProcessor()

    let firstResult = processor.handle(
      args: [
        "name": "Morgan Lee",
        "company": "OpenClaw",
        "source_type": "badge",
        "confidence": 0.88
      ],
      now: Date(timeIntervalSince1970: 0)
    )

    guard case .accepted = firstResult else {
      return XCTFail("Expected first extraction to be accepted, got \(firstResult)")
    }

    let secondResult = processor.handle(
      args: [
        "name": "Morgan Lee",
        "company": "OpenClaw",
        "source_type": "badge",
        "confidence": 0.91
      ],
      now: Date(timeIntervalSince1970: 5)
    )

    XCTAssertEqual(secondResult, .ignoredDuplicate)
  }

  private func makeProcessor() -> ConferenceExtractionProcessor {
    ConferenceExtractionProcessor(
      config: ConferenceModeConfig(
        enabled: true,
        acceptedConfidenceMin: 0.70,
        reviewConfidenceMin: 0.50,
        duplicateCooldown: 10
      )
    )
  }
}
